pkg/kubelet/cm/deviceplugin/endpoint_test.go

```diff
@@ -54,22 +54,60 @@ func TestRun(t *testing.T) {
 		{ID: "AThirdDeviceId", Health: pluginapi.Healthy},
 	}
 
-	p, e := esetup(t, devs, socket, "mock", func(n string, a, u, r []pluginapi.Device) {
-		require.Len(t, a, 1)
-		require.Len(t, u, 1)
-		require.Len(t, r, 1)
-
-		require.Equal(t, a[0].ID, updated[1].ID)
-
-		require.Equal(t, u[0].ID, updated[0].ID)
-		require.Equal(t, u[0].Health, updated[0].Health)
+	callbackCount := 0
+	callbackChan := make(chan int)
+	callback := func(n string, a, u, r []pluginapi.Device) {
+		// Should be called twice:
+		// one for plugin registration, one for plugin update.
+		if callbackCount > 2 {
+			t.FailNow()
+		}
+
+		// Check plugin registration
+		if callbackCount == 0 {
+			require.Len(t, a, 2)
+			require.Len(t, u, 0)
+			require.Len(t, r, 0)
+		}
+
+		// Check plugin update
+		if callbackCount == 1 {
+			require.Len(t, a, 1)
+			require.Len(t, u, 1)
+			require.Len(t, r, 1)
+
+			require.Equal(t, a[0].ID, updated[1].ID)
+			require.Equal(t, u[0].ID, updated[0].ID)
+			require.Equal(t, u[0].Health, updated[0].Health)
+			require.Equal(t, r[0].ID, devs[1].ID)
+		}
+
+		callbackCount++
+		callbackChan <- callbackCount
+	}
 
-		require.Equal(t, r[0].ID, devs[1].ID)
-	})
+	p, e := esetup(t, devs, socket, "mock", callback)
 	defer ecleanup(t, p, e)
 
 	go e.run()
+	// Wait for the first callback to be issued.
+	select {
+	case <-callbackChan:
+		break
+	case <-time.After(3 * time.Second):
+		t.FailNow()
+	}
```

Comment:
Suggest to just do unconditional wait with <-callbackChan for simplicity.

Reply:
If you do that tests will hang indefinitely when introducing a breaking change.
Suggest to keep that pattern

Reply:
If we really want to keep this timeout checking, lets use wait.ForeverTestTimeout as most other k8s code does instead of introducing our own test upper bound. TBH, I think we can easily control test timeout with the 'timeout' flag of 'go test' and I am pretty sure any auto test framework will set some test timeout to prevent a single test from running too long. At lower level, I would rather we just keep things simple.

Reply:
```
$ grep -R 'case <-time.After(' | wc -l
41
$ grep -R 'case <-time.After(wait.ForeverTestTimeout)' | wc -l
22
$ # Count number of files this appears in
$ grep -R 'case <-time.After(wait.ForeverTestTimeout)' | cut -d ':' -f 1 | sort | uniq | wc -l
8
$ grep -R 'case <-time.After(' | cut -d ':' -f 1 | sort | uniq | wc -l
22
```
wait.ForeverTestTimeout doesn't seem to be used that much across the code base. The pattern here is actually more frequent.

> TBH, I think we can easily control test timeout with the 'timeout' flag of 'go test' and I am pretty sure any auto test framework will set some test timeout to prevent a single test from running too long

We have a lot of sleeps in our test code, doing that in the current code means you actually have to compute the timeout bound before launching your tests.

Furthermore it makes more sense to have an explicit timeout rather than assume the implicit use of a timeout done by the test framework. 

Also note that using the test timeout framework pollutes your backtrace and makes the error less explicit.

Reply:
@jiayingz agreed with you, and `control test timeout with the 'timeout' flag of 'go test' ` looks good to me:
```sh
$ go test -v --run TestRun -timeout 3s
=== RUN   TestRun
2018/01/12 20:04:35 Starting to serve on /tmp/mock.sock
2018/01/12 20:04:35 ListAndWatch
E0112 20:04:35.758104   11765 endpoint.go:145] Device ADeviceId is now Unhealthy
E0112 20:04:35.758851   11765 endpoint.go:160] Device AnotherDeviceId was deleted
--- PASS: TestRun (0.00s)
PASS
ok  	k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin	0.039s

```
@RenaudWasTaken 

> Furthermore it makes more sense to have an explicit timeout rather than assume the implicit use of a timeout done by the test framework.

+1 :)